azure-kusto-data/src/arrow.rs

```diff
@@ -1,184 +1,146 @@
-    let trimmed = seg.trim_start_matches('0');
-    if !trimmed.is_empty() {
-        trimmed.parse::<i64>().unwrap()
-    } else {
-        0
-    }
-}
-
-fn destructure_time(dur: &str) -> (i64, i64, i64) {
-    let parts = dur.split(':').collect::<Vec<_>>();
-    match parts.as_slice() {
-        [hours, minutes, seconds] => (
-            parse_segment(hours),
-            parse_segment(minutes),
-            parse_segment(seconds),
-        ),
-        _ => (0, 0, 0),
-    }
-}
+use azure_core::error::{ErrorKind, ResultExt};
 
-/// The timespan format Kusto returns is 'd.hh:mm:ss.ssssss' or 'hh:mm:ss.ssssss' or 'hh:mm:ss'
-/// Kusto also stores fractions in ticks: 1 tick = 100 ns
-pub fn string_to_duration_i64(dur: Option<&str>) -> Option<i64> {
-    let dur = dur?;
-    let factor = if dur.starts_with('-') { -1 } else { 1 };
-    let parts: Vec<&str> = dur.trim_start_matches('-').split('.').collect();
-    let ns = match parts.as_slice() {
-        [days, hours, ticks] => {
-            let days_ = parse_segment(days);
-            let ticks_ = parse_segment(ticks);
-            let (hours, minutes, seconds) = destructure_time(hours);
-            to_nanoseconds(days_, hours, minutes, seconds, ticks_)
-        }
-        [first, ticks] => {
-            let ticks_ = parse_segment(ticks);
-            let (hours, minutes, seconds) = destructure_time(first);
-            to_nanoseconds(0, hours, minutes, seconds, ticks_)
-        }
-        [one] => {
-            let (hours, minutes, seconds) = destructure_time(one);
-            to_nanoseconds(0, hours, minutes, seconds, 0)
-        }
-        _ => 0,
-    };
-    Some(factor * ns)
-}
+use crate::error::Result;
+use crate::operations::query::*;
+use crate::operations::types::{KustoDateTime, KustoDuration};
 
-fn convert_array_string(values: Vec<serde_json::Value>) -> ArrayRef {
-    let strings: Vec<Option<String>> =
-        serde_json::from_value(serde_json::Value::Array(values)).unwrap();
+fn convert_array_string(values: Vec<serde_json::Value>) -> Result<ArrayRef> {
+    let strings: Vec<Option<String>> = serde_json::from_value(serde_json::Value::Array(values))?;
     let strings: Vec<Option<&str>> = strings.iter().map(|opt| opt.as_deref()).collect();
-    Arc::new(StringArray::from(strings))
+    Ok(Arc::new(StringArray::from(strings)))
 }
 
-// TODO provide a safe variant for datetime conversions (chrono panics)
-fn convert_array_datetime_unsafe(values: Vec<serde_json::Value>) -> ArrayRef {
-    let strings: Vec<Option<String>> =
-        serde_json::from_value(serde_json::Value::Array(values)).unwrap();
-    let strings: Vec<Option<&str>> = strings.iter().map(|opt| opt.as_deref()).collect();
-    let string_array: ArrayRef = Arc::new(StringArray::from(strings));
-    cast(
-        &string_array,
-        &DataType::Timestamp(TimeUnit::Nanosecond, None),
-    )
-    .unwrap()
+fn convert_array_datetime(values: Vec<serde_json::Value>) -> Result<ArrayRef> {
```

Comment:
This was called unsafe, since the internal conversion in arrow fails (panics) without returning an error object when the datetime is lower or higher then some thresholds. actually the same as Pandas has. Should be adjust the test data to cover that case? For performance reasonsons, I wanted to provide a variant which "hopes for the best" and another one which actually checks the string values to make sure we are always safe.

Reply:
With the new conversion it just seem to return `None` on dates outside its range.
If you find a malicious input that causes a panic we can rethink.